From 0203bdde00fb83fef4d4cfc6260d8535ed938e10 Mon Sep 17 00:00:00 2001
From: Stefan Haller <stefan@haller-berlin.de>
Date: Mon, 8 Nov 2010 10:35:34 +0100
Subject: [PATCH 06/21] gitk: Add "First parent" checkbox

Sometimes it's desirable to see what changes were introduced by a
merge commit, rather than how conflicts were resolved. This adds
a checkbox which, when turned on, makes gitk show the equivalent
of "git show --first-parent <commit>" for merge commits.

Signed-off-by: Stefan Haller <stefan@haller-berlin.de>
---
 gitk-git/gitk | 25 ++++++++++++++++++++++---
 1 file changed, 22 insertions(+), 3 deletions(-)

diff --git a/gitk-git/gitk b/gitk-git/gitk
index 5747b56ed4..5f5e3ac397 100755
--- a/gitk-git/gitk
+++ b/gitk-git/gitk
@@ -2665,6 +2665,10 @@ proc makewindow {} {
     trace add variable worddiff write changeworddiff
     pack .bleft.mid.worddiff -side left -padx 5
 
+    ttk::checkbutton .bleft.mid.firstparent -text [mc "First parent"] \
+	-command changefirstparent -variable firstparent
+    pack .bleft.mid.firstparent -side left -padx 5
+
     set ctext .bleft.bottom.ctext
     text $ctext -background $bgcolor -foreground $fgcolor \
         -state disabled -undo 0 -font textfont \
@@ -7498,6 +7502,7 @@ proc selectline {l isnew {desired_loc {}} {switch_to_patch 0}} {
     global targetrow targetid lastscrollrows
     global autocopy autoselect autosellen jump_to_here
     global vinlinediff
+    global firstparent
 
     unset -nocomplain pending_select
     unset -nocomplain pending_select_file
@@ -7651,7 +7656,7 @@ proc selectline {l isnew {desired_loc {}} {switch_to_patch 0}} {
         gettree $id
     } elseif {$vinlinediff($curview) == 1} {
         showinlinediff $id
-    } elseif {[llength $olds] <= 1} {
+    } elseif {[llength $olds] <= 1 || $firstparent} {
         startdiff $id
     } else {
         mergediff $id
@@ -8096,7 +8101,7 @@ proc diffcmd {ids flags} {
 proc gettreediffs {ids} {
     global treediff treepending limitdiffs vfilelimit curview
 
-    set cmd [diffcmd $ids {--no-commit-id}]
+    set cmd [diffcmd $ids {--no-commit-id -m --first-parent}]
     if {$limitdiffs && $vfilelimit($curview) ne {}} {
             set cmd [concat $cmd -- $vfilelimit($curview)]
     }
@@ -8193,21 +8198,34 @@ proc initblobdiffvars {} {
     set diffseehere -1
 }
 
+proc changefirstparent {} {
+    global treediffs
+    catch {unset treediffs}
+
+    reselectline
+}
+
 proc getblobdiffs {ids} {
     global blobdifffd diffids env
     global treediffs
     global diffcontext
     global ignorespace
     global worddiff
+    global firstparent
     global limitdiffs vfilelimit curview
 
-    set cmd [diffcmd $ids "-p --textconv --submodule -C --cc --no-commit-id -U$diffcontext"]
+    set cmd [diffcmd $ids "-p --textconv --submodule -C --no-commit-id -U$diffcontext"]
     if {$ignorespace} {
         append cmd " -w"
     }
     if {$worddiff ne [mc "Line diff"]} {
         append cmd " --word-diff=porcelain"
     }
+    if {$firstparent} {
+        append cmd " -m --first-parent"
+    } else {
+        append cmd " --cc"
+    }
     if {$limitdiffs && $vfilelimit($curview) ne {}} {
         set cmd [concat $cmd -- $vfilelimit($curview)]
     }
@@ -12590,6 +12608,7 @@ set diffcontext 3
 set mergecolors {red blue "#00ff00" purple brown "#009090" magenta "#808000" "#009000" "#ff0080" cyan "#b07070" "#70b0f0" "#70f0b0" "#f0b070" "#ff70b0"}
 set ignorespace 0
 set worddiff ""
+set firstparent 0
 set markbgcolor "#e0e0ff"
 
 set headbgcolor "#00ff00"
-- 
2.51.0.188.g7229e14d4b.dirty

